src/Native/FactorizationMachineNative/FactorizationMachineCoreFMA.cpp

```diff
@@ -0,0 +1,179 @@
+    _tmp = _mm256_add_ps(_y, _mm256_permute2f128_ps(_y, _y, 1));
+    _tmp = _mm256_hadd_ps(_tmp, _tmp);
+    _y = _mm256_hadd_ps(_tmp, _tmp);
+    _mm_store_ss(&latentResponse, _mm256_castps256_ps128(_y));
+    *response = linearResponse + latentResponse;
+
+}
+
+// Calculate the stochastic gradient and update the model.
+// The /*const*/ comment on the parameters of the function means that their values should not get altered by this function.
+EXPORT_API(void) CalculateGradientAndUpdateNativeFMA(float lambdaLinear, float lambdaLatent, float learningRate, int fieldCount, int latentDim, float weight, int count,
+    _In_ int* /*const*/ fieldIndices, _In_ int* /*const*/ featureIndices, _In_ float* /*const*/ featureValues, _In_ float* /*const*/ latentSum, float slope,
+    _Inout_ float* linearWeights, _Inout_ float* latentWeights, _Inout_ float* linearAccumulatedSquaredGrads, _Inout_ float* latentAccumulatedSquaredGrads)
+{
+    const int m = fieldCount;
+    const int d = latentDim;
+    const int c = count;
+    const int * pf = fieldIndices;
+    const int * pi = featureIndices;
+    const float * px = featureValues;
+    const float * pq = latentSum;
+    float * pw = linearWeights;
+    float * pv = latentWeights;
+    float * phw = linearAccumulatedSquaredGrads;
+    float * phv = latentAccumulatedSquaredGrads;
+
+    const __m256 _wei = _mm256_set1_ps(weight);
+    const __m256 _s = _mm256_set1_ps(slope);
+    const __m256 _lr = _mm256_set1_ps(learningRate);
+    const __m256 _lambdav = _mm256_set1_ps(lambdaLatent);
+
+    for (int i = 0; i < count; i++)
+    {
+        const int f = pf[i];
+        const int j = pi[i];
+
+        // Calculate gradient of linear term w_j.
+        float g = weight * (lambdaLinear * pw[j] + slope * px[i]);
+
+        // Accumulate the gradient of the linear term.
+        phw[j] += g * g;
+
+        // Perform ADAGRAD update rule to adjust linear term.
+        pw[j] -= learningRate / sqrt(phw[j]) * g;
+
+        // Update latent term, v_j,f', f'=1,...,m.
+        const __m256 _x = _mm256_broadcast_ss(px + i);
+        for (int fprime = 0; fprime < m; fprime++)
+        {
+            float * vjfprime = pv + j * m * d + fprime * d;
+            float * hvjfprime = phv + j * m * d + fprime * d;
+            const float * qfprimef = pq + fprime * m * d + f * d;
+            const __m256 _sx = _mm256_mul_ps(_s, _x);
+
+            for (int k = 0; k + 8 <= d; k += 8)
+            {
+                __m256 _v = _mm256_load_ps(vjfprime + k);
+                __m256 _q = _mm256_load_ps(qfprimef + k);
+
+                // Calculate L2-norm regularization's gradient.
+                __m256 _g = _mm256_mul_ps(_lambdav, _v);
+
+                // Calculate loss function's gradient.
+                if (fprime != f)
+                   _g = _mm256_fmadd_ps(_sx, _q, _g);
+                else
+                   _g = _mm256_fmadd_ps(_sx, _mm256_sub_ps(_q, _mm256_mul_ps(_v, _x)), _g);
```

Comment:
Not sure if we can avoid this potential branch. As least, I failed a long time ago. :)